ompi/mca/io/base/io_base_file_select.c

```diff
@@ -242,6 +236,14 @@ int mca_io_base_file_select(ompi_file_t *file,
         }
 
     }
+
+    /* Save the pointers of the selected module on the ompi_file_t */
+
+    file->f_io_version = selected.ai_version;
+    file->f_io_selected_component = selected.ai_component;
+    file->f_io_selected_module = selected.ai_module;
+    file->f_io_selected_data = selected.ai_module_data;
+
```

Comment:
Maybe I am missing something, but I do not see how moving these four lines down to after the if-statement changes anything. The if-statement executes some ompio specific initialization of the fs, fcoll, fbtl and sharedfp structures, but does not do anything with the selected module ?

Reply:
It will return to here:

https://github.com/open-mpi/ompi/blob/6d237e85d730ed946c9f45fcd3e19b78a243203e/ompi/file/file.c#L135

and try and free the file object. At this point it will eventually go down and try to defreference a NULL communicator based on one of these initializations. Delaying these initializations until after all components are opened fixes it.

I don't have the exact stack trace handy, but I can reproduce and post if that helps.

Reply:
ah, ok, I see, thanks for the explanation!